### PR TITLE
Tmpdir port pathlib

### LINKED ISSUE #3985: Introduce tmp_path fixture
After the excellent brain-storming with @gaborbernat in https://github.com/pytest-dev/pytest/pull/2230#issuecomment-421454260, I came to realize we already depend now on `pathlib2`, so we could already introduce a new `tmp_path` fixture which works exactly like `tmpdir`, but returns a `pathlib.Path` instance. 

Even if we don't ever adopt the `pathlib`-switch discussed in #2230, still seems like a good idea given that `pathlib` is part of the standard library, so makes perfect sense.

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this pr is to port tmpdir over to pathlib

the following steps have to be implemented

- [x] port the internals to pathlib and leave only a py.path shim
- [x] implement a basic make numbered dir
- [x] add a tmp_path fixture
- [x] reimplement the clean-up/folder retention
- [x] implement forced delete for rmdir (take from https://docs.python.org/3/library/shutil.html#rmtree-example)

Fix #3985

### CODE REVIEW

File: src/_pytest/tmpdir.py [hunk 1]
```diff
@@ -2,21 +2,116 @@
 from __future__ import absolute_import, division, print_function
 
 import re
+from six.moves import reduce
```
Comment (asottile) [inline src/_pytest/tmpdir.py]: in py27+ you can consistently import this `from functools import reduce` -- would be a regression of https://github.com/pytest-dev/pytest/pull/3609

File: src/_pytest/tmpdir.py [hunk 2]
```diff
@@ -33,46 +128,13 @@ def mktemp(self, basename, numbered=True):
...
-                else:
-                    rootdir = temproot
-                rootdir.ensure(dir=1)
-                basetemp = py.path.local.make_numbered_dir(
-                    prefix="pytest-", rootdir=rootdir
-                )
-            self._basetemp = t = basetemp.realpath()
-            self.trace("new basetemp", t)
-            return t
+        return py.path.local(self.tmppath_factory.getbasetemp().resolve())
```
Comment (asottile) [inline src/_pytest/tmpdir.py]: ah ok, so it still returns a `py.path` for now?

File: src/_pytest/tmpdir.py [hunk 3]
```diff
@@ -2,21 +2,116 @@
...
+        # try up to 10 times to create the folder
+        max_existing = reduce(
+            max,
+            (
+                parse_num(x)
+                for x in root.iterdir()
+                if x.name.lower().startswith(l_prefix)
+            ),
+            -1,
+        )
```
Comment (asottile) [inline src/_pytest/tmpdir.py]: this code was a major head scratcher for me, maybe refactor it to not use `reduce`?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -20,7 +20,9 @@ repos:
     -   id: check-yaml
     -   id: debug-statements
         exclude: _pytest/debugging.py
+        language_version: python3
     -   id: flake8
+        language_version: python3
```
Comment (asottile) [inline .pre-commit-config.yaml]: hmmm why was this necessary?

File: src/_pytest/tmpdir.py (hunk 2 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: the shim and the old api must stay compatible

File: src/_pytest/tmpdir.py (hunk 3 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: there is not `max(iterable, default=...)` on python2

File: .pre-commit-config.yaml (hunk 4 above)
Comment (RonnyPfannschmidt) [reply .pre-commit-config.yaml]: it used python2 on my box and crashed the flake8 builds on incompatible syntax

File: src/_pytest/tmpdir.py (hunk 2 above)
Comment (asottile) [reply src/_pytest/tmpdir.py]: makes sense, another approach might be to write an adapter class which has the interface of `py.path._Path` but keeps a `pathlib` object internally -- I suspect we'll want something like this anyway to make a path forward for methods like `as_cwd()` which have no equivalent in `pathlib`

File: .pre-commit-config.yaml (hunk 4 above)
Comment (RonnyPfannschmidt) [reply .pre-commit-config.yaml]: we do on a few scripts (like the one to make releases
the thing is - when pre-commit is installed using python3 the issue doesn't seem to happen, when it is installed with python2, it crashed

File: src/_pytest/tmpdir.py (hunk 3 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: i extracted it a bit - the for loop is slower due to executing on the interpreter, impending push after i have a first iteration of folder cleanup

File: src/_pytest/tmpdir.py (hunk 2 above)
Comment (asottile) [reply src/_pytest/tmpdir.py]: I'd be a little sad to see some of the apis go.  I for one use:

```python
with tmpdir.join('somepath').ensure().as_cwd():
    ...
```

quite a lot

Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: ensure is going to go away as well - i wonder where you use that patter, since its a bit pathological in common usage and i consider it a anti-pattern

Comment (asottile) [reply src/_pytest/tmpdir.py]: most of my tests, actually 😆 

```
$ all-repos-grep -l -E '\.(as_cwd|ensure(|_dir))\(\)'
repos/asottile/all-repos:tests/push/github_pull_request_test.py
repos/asottile/all-repos-depends:tests/conftest.py
repos/asottile/all-repos-depends:tests/depends_test.py
repos/asottile/all-repos-depends:tests/lang/python_test.py
repos/asottile/all-repos-depends:tests/packages_test.py
repos/asottile/aspy.refactor_imports:tests/classify_test.py
repos/asottile/aspy.refactor_imports:tests/conftest.py
repos/asottile/aspy.refactor_imports:tests/sort_test.py
repos/asottile/git-code-debt:tests/server/app_test.py
repos/asottile/no-manylinux1:tests/integration_test.py
repos/asottile/reorder_python_imports:tests/reorder_python_imports_test.py
repos/asottile/rubyvenv:tests/rubyvenv_test.py
repos/asottile/seed-isort-config:tests/seed_isort_config_test.py
repos/asottile/yesqa:tests/yesqa_test.py
repos/pre-commit/pre-commit:tests/commands/run_test.py
repos/pre-commit/pre-commit:tests/git_test.py
repos/pre-commit/pre-commit:tests/languages/pygrep_test.py
repos/pre-commit/pre-commit:tests/parse_shebang_test.py
repos/pre-commit/pre-commit:tests/repository_test.py
repos/pre-commit/pre-commit:tests/staged_files_only_test.py
repos/pre-commit/pre-commit-hooks:tests/autopep8_wrapper_test.py
repos/pre-commit/pre-commit-hooks:tests/check_added_large_files_test.py
repos/pre-commit/pre-commit-hooks:tests/check_case_conflict_test.py
repos/pre-commit/pre-commit-hooks:tests/check_merge_conflict_test.py
repos/pre-commit/pre-commit-hooks:tests/check_symlinks_test.py
repos/pre-commit/pre-commit-hooks:tests/check_vcs_permalinks_test.py
repos/pre-commit/pre-commit-hooks:tests/forbid_new_submodules_test.py
repos/pre-commit/pre-commit-hooks:tests/no_commit_to_branch_test.py
repos/pre-commit/pre-commit-mirror-maker:tests/make_repo_test.py
repos/sass/libsass-python:sasstests.py
```

Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: @asottile should i fix them?

Comment (asottile) [reply src/_pytest/tmpdir.py]: ehhhhhh I can fix them if the apis go away

**Comment by RonnyPfannschmidt** [comment]
just a note on the lack of progress, i caught a cold

**Comment by nicoddemus** [comment]
Sorry to hear, get better soon!

File: src/_pytest/tmpdir.py [hunk 5]
```diff
@@ -2,21 +2,222 @@
...
+    paths = find_prefixed(root, prefix)
+    paths, paths2 = itertools.tee(paths)
+    numbers = map(parse_num, extract_suffixees(paths2, prefix))
+    for path, number in zip(paths, numbers):
+        if number <= max_delete and is_deletable(path, consider_lock_dead_after):
+            delete_a_numbered_dir(path)
+
+
+def make_numbered_dir_with_cleanup(root, prefix, keep, consider_lock_dead_after):
+    for i in range(10):
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: Why the loop?

Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: this is part of the currency preparation

creating a folder may fail at the first level

a the second level in the timeframe between creating a folder and creating the lock cleanup might rename the folder

im toying with some ideas to limit the need for the loops but for practical purposes i am following the implementation of `py.path.local.make_numbered_dir` closely in that regard

**Comment by RonnyPfannschmidt** [comment]
removing the wip, a few more exception handlers/ignorers might be needed

File: src/_pytest/tmpdir.py [hunk 6]
```diff
@@ -2,21 +2,240 @@
...
+def delete_a_numbered_dir(path):
+    create_cleanup_lock(path)
+    parent = path.parent
+
+    garbage = parent.joinpath("garbage-{}".format(uuid.uuid4()))
+    path.rename(garbage)
+    shutil.rmtree(str(garbage))
+
+
+def ensure_deletable(path, consider_lock_dead_after):
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: Perhaps this should be `is_deletable` given it returns a bool?

File: src/_pytest/tmpdir.py [hunk 7]
```diff
@@ -2,21 +2,240 @@
...
+    for path in cleanup_candidates(root, prefix, keep):
+        try_cleanup(path, consider_lock_dead_after)
+    known_garbage = list(root.glob("garbage-*"))
+    for path in known_garbage:
+        try_cleanup(path, consider_lock_dead_after)
 
 
+def make_numbered_dir_with_cleanup(root, prefix, keep, lock_timeout):
+    for i in range(10):
+        try:
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: this try/except seems unnecessary given that it only re-raises any caught exception

File: src/_pytest/tmpdir.py [hunk 8]
```diff
@@ -2,21 +2,240 @@
...
+                basetemp.mkdir()
+            else:
+                temproot = Path(tempfile.gettempdir())
+                user = get_user() or "unknown"
+                # use a sub-directory in the temproot to speed-up
+                # make_numbered_dir() call
+                rootdir = temproot.joinpath("pytest-of-{}".format(user))
+                rootdir.mkdir(exist_ok=True)
+                basetemp = make_numbered_dir_with_cleanup(
+                    prefix="pytest-", root=rootdir, keep=3, lock_timeout=10000
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: 10000 seconds is 2.7 hours, the original make_numbered_dir from `py` [uses 172800](https://github.com/pytest-dev/py/blob/796d5a0d92f6eef6f5e28dfffa3a613c73a77b22/py/_path/local.py#L803). Is this intentional?

Also it helps to write `60 * 60 * 3` instead of `10800  # 3 hours` IMHO.

File: src/_pytest/tmpdir.py [hunk 9]
```diff
@@ -2,21 +2,240 @@
...
+            register_cleanup_lock_removal(lock_path)
+        except Exception:
+            raise
+        else:
+            consider_lock_dead_after = p.stat().st_mtime + lock_timeout
+            cleanup_numbered_dir(
+                root=root,
+                prefix=prefix,
+                keep=keep,
+                consider_lock_dead_after=consider_lock_dead_after,
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: Is this correct? We are computing a `consider_lock_dead_after` as the current time (we just created `p`) + some time in the future, and we are using this date in the future to decide if a lock file can be considered dead in `ensure_deletable`. It seems to me we should be passing the current time plus a `lock_timeout`, and if the last `mtime` of a lock is less than `current_time - lock_timeout`, the lock can be considered dead.

Or am I missing something?

File: src/_pytest/tmpdir.py [hunk 10]
```diff
@@ -2,21 +2,240 @@
...
+            )
+            return p
+
+
+@attr.s
+class TempPathFactory(object):
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: Please add docstrings to all the public functions here as they will appear in the docs.

File: src/_pytest/tmpdir.py [hunk 11]
```diff
@@ -2,21 +2,240 @@
...
 class TempdirFactory(object):
     """Factory for temporary directories under the common base temp directory.
 
     The base directory can be configured using the ``--basetemp`` option.
     """
 
-    def __init__(self, config):
-        self.config = config
-        self.trace = config.trace.get("tmpdir")
+    tmppath_factory = attr.ib()
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: We probably want to keep this private no?

File: src/_pytest/tmpdir.py [hunk 12]
```diff
@@ -129,3 +312,8 @@ def tmpdir(request, tmpdir_factory):
         name = name[:MAXVAL]
     x = tmpdir_factory.mktemp(name, numbered=True)
     return x
+
+
+@pytest.fixture
+def tmp_path(tmpdir):
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: docstring please

File: testing/test_tmpdir.py [hunk 13]
```diff
@@ -143,6 +143,7 @@ def break_getuser(monkeypatch):
         monkeypatch.delenv(envvar, raising=False)
 
 
+@pytest.mark.skip(reason="creates random tmpdirs as part of a system level test")
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]: Are this `skip` and the next one in this file temporary?

File: testing/test_tmpdir.py [hunk 14]
```diff
@@ -184,3 +186,70 @@ def test_get_user(monkeypatch):
...
+        assert not lock.exists()
+
+    def test_cleanup_keep(self, tmp_path):
+        self.test_make(tmp_path)
+        from _pytest.tmpdir import cleanup_numbered_dir
+
+        cleanup_numbered_dir(
+            root=tmp_path, prefix=self.PREFIX, keep=2, consider_lock_dead_after=0
+        )
+        a, b = tmp_path.iterdir()
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]: Seems incomplete

File: src/_pytest/tmpdir.py (hunk 6 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: its a modifying operation

File: src/_pytest/tmpdir.py (hunk 11 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/tmpdir.py]: yup

**Comment by RonnyPfannschmidt** [comment]
the basic structural details are finished, but `shutil.rmtree` is insufficient since its rather limited,
i must reimplement `rm -rf`

File: src/_pytest/pathlib.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        else:
+            return False
+
+
+def try_cleanup(path, consider_lock_dead_if_created_before):
+    if ensure_deletable(path, consider_lock_dead_if_created_before):
+        delete_a_numbered_dir(path)
+
+
+def cleanup_candidates(root, prefix, keep):
```
Comment (nicoddemus) [inline src/_pytest/pathlib.py]: This will not actually keep the number of directories, but will remove all directories with prefix smaller than the maximum one found - `keep`. For example:

```
fun-10
fun-11
fun-12
fun-16
fun-20
```

With `keep=3`, only `fun-20` will be left in the directory. `keep=3` could be considered "leave the last 3 folders", in which case `fun-12`, `fun-16` and `fun-20` should be kept.

I don't mind either way, just mentioning this to make sure we are getting the behavior we want (and we should document it to be clear as well).

File: src/_pytest/tmpdir.py (hunk 6 above)
Comment (nicoddemus) [reply src/_pytest/tmpdir.py]: Oh right, my bad

File: src/_pytest/pathlib.py (hunk 15 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/pathlib.py]: @nicoddemus its the current behaviour

Comment (nicoddemus) [reply src/_pytest/pathlib.py]: I see, thanks. Let's make sure to document it later then. 👍

**Comment by RonnyPfannschmidt** [comment]
i'll have to leave out forced delete, the python stdlib is too broken to do it reliably

**Comment by RonnyPfannschmidt** [comment]
aka this will be a followup

File: src/_pytest/deprecated.py [hunk 16]
```diff
@@ -109,3 +109,8 @@
 PYTEST_NAMESPACE = RemovedInPytest4Warning(
     "pytest_namespace is deprecated and will be removed soon"
 )
+
+PYTEST_ENSURETEMP = RemovedInPytest4Warning(
```
Comment (nicoddemus) [inline src/_pytest/deprecated.py]: We should add a `deprecation` changelog entry for that. 👍

File: doc/en/tmpdir.rst [hunk 17]
```diff
@@ -5,6 +5,51 @@
...
+
+
+
+The ``tmp_path_facotry`` fixture
+------------------------------
+
+.. versionadded:: 3.9
+
+
+The ``tmp_path_facotry`` is a session-scoped fixture which can be used
```
Comment (nicoddemus) [inline doc/en/tmpdir.rst]: I think it is also important to mention that it returns `pathlib` objects.

File: doc/en/tmpdir.rst [hunk 18]
```diff
@@ -5,6 +5,51 @@
...
+.. versionadded:: 3.9
+
+
+You can use the ``tmpdir`` fixture which will
+provide a temporary directory unique to the test invocation,
+created in the `base temporary directory`_.
+
+``tmpdir`` is a `pathlib/pathlib2.Path`_ object. Here is an example test usage::
+
+    # content of test_tmp_path.py
```
Comment (nicoddemus) [inline doc/en/tmpdir.rst]: I believe it is better to use `.. code-block:: python` so that blaken-docs can do its thing.

File: src/_pytest/tmpdir.py [hunk 19]
```diff
@@ -1,22 +1,86 @@
...
-class TempdirFactory(object):
+
+@attr.s
+class TempPathFactory(object):
     """Factory for temporary directories under the common base temp directory.
 
-    The base directory can be configured using the ``--basetemp`` option.
+    The base directory can be configured using the ``--basetemp`` option."""
+
+    given_basetemp = attr.ib()
```
Comment (nicoddemus) [inline src/_pytest/tmpdir.py]: This is still public, I think it should really be private

File: src/_pytest/tmpdir.py [hunk 20]
```diff
@@ -2,21 +2,74 @@
...
+    ensure_reset_dir,
+)
+
+
+@attr.s
+class TempPathFactory(object):
+    """docstring for ClassName"""
+
+    given_basetemp = attr.ib()
+    trace = attr.ib()
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply src/_pytest/tmpdir.py]: This is still public, I think it should really be private

File: .pre-commit-config.yaml [hunk 21]
```diff
@@ -43,6 +43,6 @@ repos:
     -   id: changelogs-rst
         name: changelog filenames
         language: fail
-        entry: 'changelog files must be named ####.(feature|bugfix|doc|removal|vendor|trivial).rst'
+        entry: 'changelog files must be named ####.(feature|bugfix|doc|deprecation|removal|vendor|trivial).rst'
```
Comment (nicoddemus) [inline .pre-commit-config.yaml]: Oops my bad, missed this one

Comment (RonnyPfannschmidt) [reply .pre-commit-config.yaml]: i just noticed it due to a typo in the filename, so it was only a display error, not a technical error ^^

File: doc/en/tmpdir.rst (hunk 17 above)
Comment (asottile) [reply doc/en/tmpdir.rst]: Also, "facotry" :D

Comment (nicoddemus) [reply doc/en/tmpdir.rst]: Oops good catch

Comment (nicoddemus) [reply doc/en/tmpdir.rst]: Hmm I would fix this myself but the "edit" button is disabled for me.

**Comment by nicoddemus** [comment]
🎉
